riscv/src/bootloader.rs

```diff
@@ -9,6 +9,47 @@ pub const PAGE_SIZE_BYTES_LOG: usize = 10;
 // 32-Bit architecture -> 4 bytes per word
 pub const BYTES_PER_WORD: usize = 4;
 
+pub fn bootloader_preamble() -> String {
+    let mut preamble = r#"
+    // ============== bootloader-specific instructions =======================
+    // Write-once memory
+    let BOOTLOADER_INPUT_ADDRESS = |i| i;
+    let bootloader_input_value;
+    // Loads a value. If the cell is empty, the prover can choose a value.
+    instr load_bootloader_input X -> Y { {X, Y} in {BOOTLOADER_INPUT_ADDRESS, bootloader_input_value} }
+
+    // HACK: We should not use these columns & constraints and instead be able to use X.
+    // However, this seems to currently violate some constraints, because X is neither
+    // an input nor an output of the jump_to_bootloader_input_if_nonzero instruction.
+    col witness tmp_addr;
+    col witness tmp_addrInv;
+    col witness tmp_addrIsZero;
+    tmp_addrIsZero = 1 - tmp_addr * tmp_addrInv;
+    tmp_addrIsZero * tmp_addr = 0;
+    tmp_addrIsZero * (1 - tmp_addrIsZero) = 0;
+    (1 - instr_jump_to_bootloader_input_if_nonzero) * tmp_addr = 0;
+
+    // Sets the PC to the bootloader input at the provided index if it is nonzero
+    instr jump_to_bootloader_input_if_nonzero Y {
+        // Atomically reads the bootloader input at index Y and jumps to it if it is nonzero
+        // (otherwise, the PC is incremented by 1).
+        {Y, tmp_addr} in {BOOTLOADER_INPUT_ADDRESS, bootloader_input_value},
+        pc' = (1 - tmp_addrIsZero) * tmp_addr + tmp_addrIsZero * (pc + 1)
+    }
+
+    // Expose initial register values as public outputs
+"#.to_string();
+
+    for (i, reg) in REGISTER_NAMES.iter().enumerate() {
+        let reg = reg.strip_prefix("main.").unwrap();
+        preamble.push_str(&format!(
+            "    public initial_{reg} = bootloader_input_value({i});\n"
```

Comment:
where are these used?

Reply:
There are not used, this just exposes some of the bootloader inputs as public, so that the verifier can run additional checks (i.e., verifier that the prover started with the correct initial register values).

In theory, the language supports to reference these values even in identities, but I think this is not fully implemented and not needed here. So the names are just for convenience.

Reply:
Ah I see